File: sdk/python/inkbox/identities/resources/identities.py
"""
inkbox/identities/resources/identities.py

Identity CRUD. Mailbox and tunnel are provisioned atomically by
``create()``; there is no standalone mailbox / tunnel create surface.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Literal
from uuid import UUID

from inkbox.exceptions import InkboxAPIError
from inkbox.identities.exceptions import map_identity_conflict_error
from inkbox.identities.types import (
    _UNSET,
    AgentIdentitySummary,
    IdentityMailboxCreateOptions,
    IdentityPhoneNumberCreateOptions,
    IdentityTunnelCreateOptions,
    _AgentIdentityData,
    vault_secret_ids_to_wire,
)
from inkbox.imessage.types import _validate_idempotency_key

if TYPE_CHECKING:
    from inkbox._http import HttpTransport


class IdentitiesResource:
    def __init__(self, http: HttpTransport) -> None:
        self._http = http

    def create(
        self,
        *,
        agent_handle: str,
        display_name: str | None = None,
        description: Any = _UNSET,
        imessage_enabled: bool | None = None,
        contact_sharing_enabled: bool | None = None,
        claim_imessage_number: Literal[True] | None = None,
        mailbox: IdentityMailboxCreateOptions | None = None,
        tunnel: IdentityTunnelCreateOptions | None = None,
        phone_number: IdentityPhoneNumberCreateOptions | None = None,
        vault_secret_ids: UUID | str | list[UUID | str] | None = None,
    ) -> _AgentIdentityData:
        """Create a new agent identity. Atomically provisions the
        identity's mailbox and tunnel; both are returned nested on the
        response.

        Args:
            agent_handle: Unique handle, globally unique across all orgs
                (the handle shares its namespace with tunnel names). May
                be passed with or without a leading ``@``.
            display_name: Human-readable identity name. Defaults
                server-side to ``agent_handle``.
            description: Free-form org-internal description. Pass
                ``None`` to leave the column null; omit entirely to defer
                to the server default.
            imessage_enabled: Whether the identity can use iMessage. Omit to
                defer to the server default (``False``).
            contact_sharing_enabled: Whether an attached dedicated iMessage
                line automatically shares this identity's name and optional
                avatar. Omit to defer to the server default (``True``).
            claim_imessage_number: Claim and attach a new dedicated iMessage
                line atomically. Requires
                ``imessage_enabled=True``.
            mailbox: Optional nested mailbox spec.
            tunnel: Optional nested tunnel spec (tls_mode only).
            phone_number: Optional phone-number provisioning payload.
            vault_secret_ids: Optional vault secret selection to attach.

        Returns:
            The created identity with ``mailbox`` and ``tunnel``
            populated from the atomic create response.
        """
        if claim_imessage_number is not None and claim_imessage_number is not True:
            raise ValueError("claim_imessage_number must be True when supplied")
        body: dict[str, Any] = {"agent_handle": agent_handle}
        if display_name is not None:
            body["display_name"] = display_name
        if description is not _UNSET:
            body["description"] = description
        if imessage_enabled is not None:
            body["imessage_enabled"] = imessage_enabled
        if contact_sharing_enabled is not None:
            body["contact_sharing_enabled"] = contact_sharing_enabled
        if claim_imessage_number is True:
            if imessage_enabled is not True:
                raise ValueError(
                    "claim_imessage_number requires imessage_enabled=True"
                )
        if claim_imessage_number is True:
            body["claim_imessage_number"] = True
        if mailbox is not None:
            body["mailbox"] = mailbox.to_wire()
        if tunnel is not None:
            body["tunnel"] = tunnel.to_wire()
        if phone_number is not None:
            body["phone_number"] = phone_number.to_wire()
        if vault_secret_ids is not None:
            body["vault_secret_ids"] = vault_secret_ids_to_wire(vault_secret_ids)
        try:
            data = self._http.post("/", json=body)
        except InkboxAPIError as err:
            raise map_identity_conflict_error(err) from err
        return _AgentIdentityData._from_dict(data)

    def list(self) -> list[AgentIdentitySummary]:
        """List identities, including hydrated fields when provided."""
        data = self._http.get("/")
        return [AgentIdentitySummary._from_dict(i) for i in data]

    def get(self, agent_handle: str) -> _AgentIdentityData:
        """Get an identity with its linked channels (mailbox, phone
        number, tunnel)."""
        data = self._http.get(f"/{agent_handle}")
        return _AgentIdentityData._from_dict(data)

    def update(
        self,
        agent_handle: str,
        *,
        new_handle: str | None = None,
        display_name: Any = _UNSET,
        description: Any = _UNSET,
        imessage_enabled: bool | None = None,
        contact_sharing_enabled: bool | None = None,
        imessage_number_id: UUID | str | None = _UNSET,  # type: ignore[assignment]
        claim_imessage_number: Literal[True] | None = None,
        idempotency_key: str | None = None,
        imessage_filter_mode: str | None = None,
        mail_filter_mode: str | None = None,
        phone_filter_mode: str | None = None,
    ) -> _AgentIdentityData:
        """Update an identity's handle, display name, description,
        iMessage reachability, and contact-rule filter modes.

        Only provided fields are applied; omitted fields are left
        unchanged. For ``display_name`` and ``description``, explicit
        ``None`` clears the column; omitting the keyword argument leaves
        it untouched (distinguished via an internal ``_UNSET`` sentinel).

        Args:
            agent_handle: Current handle of the identity to update.
            new_handle: New handle value.
            display_name: New display name, or ``None`` to clear.
            description: New description, or ``None`` to clear.
            imessage_enabled: Toggle iMessage reachability.
            contact_sharing_enabled: Toggle automatic name and optional photo
                sharing for an attached dedicated iMessage line.
            imessage_number_id: Attach an already-owned dedicated line by
                UUID, pass ``None`` to move back to the shared service, or
                omit to leave the current attachment unchanged.
            claim_imessage_number: Claim and attach a new dedicated iMessage
                line. Cannot be combined with
                ``imessage_number_id`` and requires ``idempotency_key``.
            idempotency_key: Stable caller-generated key for a dedicated-line
                claim. Reuse the same value when retrying the same update.
            imessage_filter_mode: ``"whitelist"`` or ``"blacklist"`` for
                iMessage contact rules (admin-only).
            mail_filter_mode: ``"whitelist"`` or ``"blacklist"`` for this
                identity's mail contact rules (admin-only).
            phone_filter_mode: ``"whitelist"`` or ``"blacklist"`` for this
                identity's phone contact rules (admin-only). The server
                rejects this with 422 when the identity has no phone number.
        """
        if claim_imessage_number is not None and claim_imessage_number is not True:
            raise ValueError("claim_imessage_number must be True when supplied")
        body: dict[str, Any] = {}
        if new_handle is not None:
            body["agent_handle"] = new_handle
        if display_name is not _UNSET:
            body["display_name"] = display_name
        if description is not _UNSET:
            body["description"] = description
        if imessage_enabled is not None:
            body["imessage_enabled"] = imessage_enabled
        if contact_sharing_enabled is not None:
            body["contact_sharing_enabled"] = contact_sharing_enabled
        if claim_imessage_number is True and imessage_number_id is not _UNSET:
            raise ValueError(
                "claim_imessage_number and imessage_number_id cannot be set together"
            )
        if claim_imessage_number is True:
            if imessage_enabled is False:
                raise ValueError(
                    "claim_imessage_number cannot be set while disabling iMessage"
                )
            if idempotency_key is None:
                raise ValueError(
                    "idempotency_key is required with claim_imessage_number"
                )
        if claim_imessage_number is True:
            body["claim_imessage_number"] = True
        if imessage_number_id is not _UNSET:
            if imessage_number_id is not None and imessage_enabled is False:
                raise ValueError(
                    "imessage_number_id cannot be set while disabling iMessage"
                )
            body["imessage_number_id"] = (
                str(imessage_number_id)
                if imessage_number_id is not None
                else None
            )
        if imessage_filter_mode is not None:
            body["imessage_filter_mode"] = imessage_filter_mode
        if mail_filter_mode is not None:
            body["mail_filter_mode"] = mail_filter_mode
        if phone_filter_mode is not None:
            body["phone_filter_mode"] = phone_filter_mode
        headers = None
        if idempotency_key is not None:
            headers = {
                "Idempotency-Key": _validate_idempotency_key(idempotency_key),
            }
        try:
            if headers is None:
                data = self._http.patch(f"/{agent_handle}", json=body)
            else:
                data = self._http.patch(
                    f"/{agent_handle}", json=body, headers=headers,
                )
        except InkboxAPIError as err:
            raise map_identity_conflict_error(err) from err
        return _AgentIdentityData._from_dict(data)

    def delete(self, agent_handle: str) -> None:
        """Delete an identity.

        Cascades: flips the linked mailbox to ``deleted``, force-finalizes
        the linked tunnel to ``deleted``, revokes any identity-scoped
        API keys, and releases any linked phone number (vendor + local).
        """
        self._http.delete(f"/{agent_handle}")

    def release_phone_number(self, agent_handle: str) -> None:
        """Release the identity's phone number (vendor + local).

        Released at the carrier; the number is not available for
        reassignment afterwards.
        """
        self._http.delete(f"/{agent_handle}/phone_number")

File: sdk/typescript/src/identities/resources/identities.ts
/**
 * inkbox-identities/resources/identities.ts
 *
 * Identity create / list / get / update / delete, plus phone-number
 * release. Mailbox and tunnel are provisioned atomically by `create()`
 * and removed by `delete()` (cascade); there is no standalone mailbox
 * or tunnel create / link surface.
 */

import { HttpTransport, InkboxAPIError, validateIdempotencyKey } from "../../_http.js";
import { mapIdentityConflictError } from "../exceptions.js";
import {
  AgentIdentitySummary,
  IdentityMailboxCreateOptions,
  IdentityPhoneNumberCreateOptions,
  IdentityTunnelCreateOptions,
  UpdateIdentityOptions,
  _AgentIdentityData,
  RawAgentIdentitySummary,
  RawAgentIdentityData,
  identityMailboxCreateOptionsToWire,
  identityPhoneNumberCreateOptionsToWire,
  identityTunnelCreateOptionsToWire,
  parseAgentIdentitySummary,
  parseAgentIdentityData,
  vaultSecretIdsToWire,
} from "../types.js";

export class IdentitiesResource {
  constructor(private readonly http: HttpTransport) {}

  /**
   * Create a new agent identity. Atomically provisions the identity's
   * mailbox and tunnel; both are returned nested on the response.
   *
   * @param options.agentHandle - Unique handle for this identity, globally
   *   unique across all orgs (the handle shares its namespace with tunnel
   *   names). May be passed with or without a leading `@`.
   * @param options.displayName - Human-readable identity name. Defaults
   *   server-side to `agentHandle`.
   * @param options.description - Free-form org-internal description.
   *   `null` leaves the column null; omit to defer to server default.
   * @param options.imessageEnabled - Whether the identity can be reached
   *   over iMessage. Omit to defer to the server
   *   default (`false`).
   * @param options.contactSharingEnabled - Whether an attached dedicated
   *   iMessage line shares the identity's name and optional avatar.
   * @param options.claimIMessageNumber - Claim a dedicated iMessage line and
   *   attach atomically. Requires `imessageEnabled: true`.
   * @param options.mailbox - Optional nested mailbox spec. Mailbox is
   *   always provisioned; this just lets the caller customize.
   * @param options.tunnel - Optional nested tunnel spec (tlsMode only).
   *   Tunnel is always provisioned; defaults to edge TLS.
   * @param options.phoneNumber - Optional phone-number provisioning payload.
   * @param options.vaultSecretIds - Optional vault secret selection to attach to the identity.
   */
  async create(options: {
    agentHandle: string;
    displayName?: string;
    description?: string | null;
    imessageEnabled?: boolean;
    contactSharingEnabled?: boolean;
    claimIMessageNumber?: true;
    mailbox?: IdentityMailboxCreateOptions;
    tunnel?: IdentityTunnelCreateOptions;
    phoneNumber?: IdentityPhoneNumberCreateOptions;
    vaultSecretIds?: string | string[] | "*" | "all";
  }): Promise<_AgentIdentityData> {
    if (options.claimIMessageNumber !== undefined && options.claimIMessageNumber !== true) {
      throw new Error("claimIMessageNumber must be true when supplied");
    }
    if (options.claimIMessageNumber === true && options.imessageEnabled !== true) {
      throw new Error("claimIMessageNumber requires imessageEnabled: true");
    }
    const body: Record<string, unknown> = { agent_handle: options.agentHandle };
    if (options.displayName !== undefined) body["display_name"] = options.displayName;
    if (options.description !== undefined) body["description"] = options.description;
    if (options.imessageEnabled !== undefined) body["imessage_enabled"] = options.imessageEnabled;
    if (options.contactSharingEnabled !== undefined) body["contact_sharing_enabled"] = options.contactSharingEnabled;
    if (options.claimIMessageNumber === true) {
      body["claim_imessage_number"] = true;
    }
    if (options.mailbox !== undefined) body["mailbox"] = identityMailboxCreateOptionsToWire(options.mailbox);
    if (options.tunnel !== undefined) body["tunnel"] = identityTunnelCreateOptionsToWire(options.tunnel);
    if (options.phoneNumber !== undefined) body["phone_number"] = identityPhoneNumberCreateOptionsToWire(options.phoneNumber);
    if (options.vaultSecretIds !== undefined) body["vault_secret_ids"] = vaultSecretIdsToWire(options.vaultSecretIds);
    try {
      const data = await this.http.post<RawAgentIdentityData>("/", body);
      return parseAgentIdentityData(data);
    } catch (err) {
      if (err instanceof InkboxAPIError) throw mapIdentityConflictError(err);
      throw err;
    }
  }

  /** List identities, preserving hydrated fields when provided. */
  async list(): Promise<AgentIdentitySummary[]> {
    const data = await this.http.get<RawAgentIdentitySummary[]>("/");
    return data.map(parseAgentIdentitySummary);
  }

  /**
   * Get an identity with its linked channels (mailbox, phone number, tunnel).
   *
   * @param agentHandle - Handle of the identity to fetch.
   */
  async get(agentHandle: string): Promise<_AgentIdentityData> {
    const data = await this.http.get<RawAgentIdentityData>(`/${agentHandle}`);
    return parseAgentIdentityData(data);
  }

  /**
   * Update an identity's handle, display name, description, iMessage
   * reachability, and/or contact-rule filter modes.
   *
   * Only provided fields are applied; omitted fields are left unchanged.
   * For `displayName` and `description`, explicit `null` clears the value
   * (sets the row column to NULL); omitting the key leaves it untouched.
   *
   * @param agentHandle - Current handle of the identity to update.
   * @param options.newHandle - New handle value.
   * @param options.displayName - New display name, or `null` to clear.
   * @param options.description - New description, or `null` to clear.
   * @param options.imessageEnabled - Toggle identity-level iMessage reachability.
   * @param options.contactSharingEnabled - Toggle automatic name and optional
   *   photo sharing for an attached dedicated iMessage line.
   * @param options.imessageNumberId - Attach an owned dedicated line, or
   *   pass `null` to return to shared service.
   * @param options.claimIMessageNumber - Claim and attach a new dedicated line.
   * @param options.idempotencyKey - Stable caller-generated key required for
   *   `claimIMessageNumber`. Reuse it after an ambiguous failure.
   * @param options.imessageFilterMode - `"whitelist"` or `"blacklist"`
   *   for iMessage contact rules (admin-only).
   * @param options.mailFilterMode - `"whitelist"` or `"blacklist"` for this
   *   identity's mail contact rules (admin-only).
   * @param options.phoneFilterMode - `"whitelist"` or `"blacklist"` for this
   *   identity's phone contact rules (admin-only). The server rejects this
   *   with 422 when the identity has no phone number.
   */
  async update(
    agentHandle: string,
    options: UpdateIdentityOptions,
  ): Promise<_AgentIdentityData> {
    if (options.claimIMessageNumber !== undefined && options.claimIMessageNumber !== true) {
      throw new Error("claimIMessageNumber must be true when supplied");
    }
    const hasNumberId = "imessageNumberId" in options;
    if (options.claimIMessageNumber === true && hasNumberId) {
      throw new Error("claimIMessageNumber and imessageNumberId cannot be set together");
    }
    if (
      options.imessageEnabled === false
      && (
        options.claimIMessageNumber === true
        || (hasNumberId && options.imessageNumberId !== null)
      )
    ) {
      throw new Error("iMessage number changes cannot be combined with disabling iMessage");
    }
    if (options.claimIMessageNumber === true) {
      if (options.idempotencyKey === undefined) {
        throw new Error("idempotencyKey is required with claimIMessageNumber");
      }
    }
    if (options.idempotencyKey !== undefined) {
      validateIdempotencyKey(options.idempotencyKey);
    }
    const body: Record<string, unknown> = {};
    if (options.newHandle !== undefined) body["agent_handle"] = options.newHandle;
    if (options.displayName !== undefined) body["display_name"] = options.displayName;
    if (options.description !== undefined) body["description"] = options.description;
    if (options.imessageEnabled !== undefined) body["imessage_enabled"] = options.imessageEnabled;
    if (options.contactSharingEnabled !== undefined) body["contact_sharing_enabled"] = options.contactSharingEnabled;
    if ("imessageNumberId" in options) body["imessage_number_id"] = options.imessageNumberId;
    if (options.claimIMessageNumber === true) {
      body["claim_imessage_number"] = true;
    }
    if (options.imessageFilterMode !== undefined) body["imessage_filter_mode"] = options.imessageFilterMode;
    if (options.mailFilterMode !== undefined) body["mail_filter_mode"] = options.mailFilterMode;
    if (options.phoneFilterMode !== undefined) body["phone_filter_mode"] = options.phoneFilterMode;
    try {
      const data = options.idempotencyKey === undefined
        ? await this.http.patch<RawAgentIdentityData>(`/${agentHandle}`, body)
        : await this.http.patch<RawAgentIdentityData>(`/${agentHandle}`, body, {
          headers: { "Idempotency-Key": options.idempotencyKey },
        });
      return parseAgentIdentityData(data);
    } catch (err) {
      if (err instanceof InkboxAPIError) throw mapIdentityConflictError(err);
      throw err;
    }
  }

  /**
   * Delete an identity.
   *
   * Cascades: flips the linked mailbox to `deleted`, force-finalizes the
   * linked tunnel to `deleted`, revokes any identity-scoped API keys, and
   * releases any linked phone number (vendor + local).
   *
   * @param agentHandle - Handle of the identity to delete.
   */
  async delete(agentHandle: string): Promise<void> {
    await this.http.delete(`/${agentHandle}`);
  }

  /**
   * Release the identity's phone number (vendor + local).
   *
   * Released at the carrier; the number is not available for
   * reassignment afterwards.
   *
   * @param agentHandle - Handle of the identity.
   */
  async releasePhoneNumber(agentHandle: string): Promise<void> {
    await this.http.delete(`/${agentHandle}/phone_number`);
  }

}
